internal/persistence/sql/relationtuples.go

```diff
@@ -6,18 +6,29 @@ package sql
 import (
 	"context"
 	"database/sql"
+	"fmt"
+	"slices"
+	"strings"
 	"time"
 
-	"github.com/ory/keto/ketoapi"
-
 	"github.com/gobuffalo/pop/v6"
 	"github.com/gofrs/uuid"
 	"github.com/ory/x/otelx"
 	"github.com/ory/x/sqlcon"
 	"github.com/pkg/errors"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/trace"
 
 	"github.com/ory/keto/internal/relationtuple"
 	"github.com/ory/keto/internal/x"
+	"github.com/ory/keto/ketoapi"
+)
+
+// Typical database limits for placeholders/bind vars are 1<<15 (32k, MySQL, SQLite) and 1<<16 (64k, PostgreSQL, CockroachDB).
+const (
+	chunkSizeInsertUUIDMappings = 15000 // two placeholders per mapping
+	chunkSizeInsertTuple        = 3000  // ten placeholders per tuple
+	chunkSizeDeleteTuple        = 100   // the database must build an expression tree for each chunk, so we must limit more agressively
```

Comment:
What exactly do you mean by "build an expression tree" here?

Reply:
The DB parses the `DELETE WHERE` query and will construct a data structure for it. That's likely not an inexpensive operation especially if optimizations are performed on that structure afterwards. SQLite actually has a limit of 1000 nodes for this structure.

Inserts are a lot simpler in this regard, and we can have bigger queries.